File: src/gearbox_sim/components/battery.py
"""Battery component for hybrid and electric drivetrains."""

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

import numpy as np

from ..core.component import DrivetrainComponent
from ..core.ports import Port, PortType, PortDirection
from ..core.constraints import KinematicConstraint


@dataclass
class BatteryParams:
    """Battery parameters.

    Attributes:
        capacity_kwh: Energy capacity [kWh]
        V_oc: Open circuit voltage [V]
        V_nom: Nominal voltage [V]
        R_int: Internal resistance [Ω]
        P_max_discharge: Maximum discharge power [W]
        P_max_charge: Maximum charge power [W]
        SOC_min: Minimum operating SOC
        SOC_max: Maximum operating SOC
        SOC_init: Initial SOC
    """

    capacity_kwh: float = 200.0
    V_oc: float = 750.0
    V_nom: float = 700.0
    R_int: float = 0.05
    P_max_discharge: float = 1_000_000.0
    P_max_charge: float = 500_000.0
    SOC_min: float = 0.3
    SOC_max: float = 0.8
    SOC_init: float = 0.6

    @property
    def Q_capacity(self) -> float:
        """Capacity in Joules."""
        return self.capacity_kwh * 3600 * 1000

    @property
    def Q_Ah(self) -> float:
        """Capacity in Amp-hours."""
        return self.Q_capacity / (self.V_nom * 3600)


# Pre-configured for CAT 793D
CAT_793D_BATTERY_PARAMS = BatteryParams(
    capacity_kwh=200.0,
    V_oc=750.0,
    V_nom=700.0,
    R_int=0.05,
    P_max_discharge=1_000_000.0,
    P_max_charge=500_000.0,
    SOC_min=0.3,
    SOC_max=0.8,
    SOC_init=0.6,
)

# Larger battery for series hybrid
SERIES_HYBRID_BATTERY_PARAMS = BatteryParams(
    capacity_kwh=400.0,
    V_oc=750.0,
    V_nom=700.0,
    R_int=0.03,
    P_max_discharge=1_500_000.0,
    P_max_charge=800_000.0,
    SOC_min=0.2,
    SOC_max=0.9,
    SOC_init=0.6,
)

# Large battery for pure EV
EV_BATTERY_PARAMS = BatteryParams(
    capacity_kwh=600.0,
    V_oc=800.0,
    V_nom=750.0,
    R_int=0.02,
    P_max_discharge=2_000_000.0,
    P_max_charge=1_000_000.0,
    SOC_min=0.1,
    SOC_max=0.95,
    SOC_init=0.8,
)


class BatteryComponent(DrivetrainComponent):
    """Battery energy storage component.

    Uses simple equivalent circuit model:
    V_terminal = V_oc - I × R_int

    Power limits are derated near SOC boundaries to protect the battery.

    Ports:
        electrical: Electrical power port (bidirectional)

    Internal states:
        SOC: State of charge [0-1]
    """

    def __init__(self, params: BatteryParams = None, name: str = "battery"):
        """Initialize battery component.

        Args:
            params: Battery parameters
            name: Component name
        """
        super().__init__(name)
        self.params = params or BatteryParams()
        self._soc = self.params.SOC_init

    @property
    def soc(self) -> float:
        """Current state of charge [0-1]."""
        return self._soc

    @soc.setter
    def soc(self, value: float) -> None:
        """Set state of charge."""
        self._soc = float(np.clip(value, 0.0, 1.0))

    @property
    def ports(self) -> Dict[str, Port]:
        return {
            "electrical": Port(
                name="electrical",
                port_type=PortType.ELECTRICAL,
                direction=PortDirection.BIDIRECTIONAL,
                description="Electrical power connection",
            )
        }

    @property
    def state_names(self) -> List[str]:
        return ["SOC"]

    def get_inertia(self, port_name: str) -> float:
        # Battery has no mechanical inertia
        return 0.0

    def get_constraints(self) -> List[KinematicConstraint]:
        return []

    def get_open_circuit_voltage(self, soc: float = None) -> float:
        """Get open circuit voltage (could be SOC-dependent).

        Args:
            soc: State of charge (uses current if None)

        Returns:
            Open circuit voltage [V]
        """
        # Simple model: constant V_oc
        # Could add SOC dependency: V_oc = f(SOC)
        return self.params.V_oc

    def get_current_from_power(self, power: float, soc: float = None) -> float:
        """Calculate current from power demand.

        Solves: P = V_oc × I - R_int × I²

        Args:
            power: Electrical power [W] (positive = discharge)
            soc: State of charge (uses current if None)

        Returns:
            Current [A] (positive = discharge)
        """
        if soc is None:
            soc = self._soc

        V_oc = self.get_open_circuit_voltage(soc)
        R = self.params.R_int

        if abs(power) < 1e-6:
            return 0.0

        # Quadratic: R × I² - V_oc × I + P = 0
        # I = (V_oc - sqrt(V_oc² - 4RP)) / (2R)
        discriminant = V_oc**2 - 4 * R * power

        if discriminant < 0:
            # Power exceeds capability - return limiting current
            return V_oc / (2 * R) if power > 0 else -V_oc / (2 * R)

        # Use solution that gives smaller magnitude current
        I = (V_oc - np.sqrt(discriminant)) / (2 * R)
        return I

    def get_terminal_voltage(self, current: float, soc: float = None) -> float:
        """Get terminal voltage at given current.

        Args:
            current: Current [A] (positive = discharge)
            soc: State of charge

        Returns:
            Terminal voltage [V]
        """
        if soc is None:
            soc = self._soc

        V_oc = self.get_open_circuit_voltage(soc)
        return V_oc - current * self.params.R_int

    def get_power_limits(self, soc: float = None) -> tuple[float, float]:
        """Get power limits with SOC-based derating.

        Args:
            soc: State of charge (uses current if None)

        Returns:
            Tuple of (P_min, P_max) [W]
            P_min is negative (max charging), P_max is positive (max discharging)
        """
        if soc is None:
            soc = self._soc

        P_discharge = self.params.P_max_discharge
        P_charge = self.params.P_max_charge

        # Derate discharge at low SOC
        if soc < self.params.SOC_min + 0.1:
            factor = max(0.0, (soc - self.params.SOC_min) / 0.1)
            P_discharge *= factor

        # Derate charge at high SOC
        if soc > self.params.SOC_max - 0.1:
            factor = max(0.0, (self.params.SOC_max - soc) / 0.1)
            P_charge *= factor

        return (-P_charge, P_discharge)

    def clip_power(self, power: float, soc: float = None) -> float:
        """Clip power to valid range.

        Args:
            power: Requested power [W]
            soc: State of charge

        Returns:
            Clipped power [W]
        """
        P_min, P_max = self.get_power_limits(soc)
        return float(np.clip(power, P_min, P_max))

    def get_soc_derivative(self, power: float, soc: float = None) -> float:
        """Calculate SOC rate of change.

        dSOC/dt = -I / Q_capacity

        Args:
            power: Electrical power [W] (positive = discharge)
            soc: Current SOC

        Returns:
            SOC rate [1/s]
        """
        if soc is None:
            soc = self._soc

        current = self.get_current_from_power(power, soc)
        # Q_capacity is in Coulombs (A·s)
        Q = self.params.Q_capacity / self.params.V_nom  # Convert to Coulombs
        return -current / Q

    def compute_torques(
        self,
        port_speeds: Dict[str, float],
        control_inputs: Dict[str, float],
        internal_states: Optional[Dict[str, float]] = None,
    ) -> Dict[str, float]:
        """Battery doesn't produce torque."""
        return {}

    def compute_state_derivatives(
        self,
        internal_states: Dict[str, float],
        port_values: Dict[str, Any],
    ) -> Dict[str, float]:
        """Compute SOC derivative.

        Args:
            internal_states: {"SOC": current_soc}
            port_values: Should contain "electrical_power" or be computed
                        from connected motor powers

        Returns:
            {"SOC": dSOC/dt}
        """
        soc = internal_states.get("SOC", self._soc)
        power = port_values.get("electrical_power", 0.0)

        d_soc = self.get_soc_derivative(power, soc)
        return {"SOC": d_soc}

    def can_provide_power(self, power: float, soc: float = None) -> bool:
        """Check if battery can provide requested power.

        Args:
            power: Requested power [W]
            soc: State of charge

        Returns:
            True if power is achievable
        """
        P_min, P_max = self.get_power_limits(soc)
        return P_min <= power <= P_max

    def get_energy_remaining(self, soc: float = None) -> float:
        """Get remaining usable energy.

        Args:
            soc: State of charge

        Returns:
            Usable energy [J]
        """
        if soc is None:
            soc = self._soc

        usable_soc = max(0.0, soc - self.params.SOC_min)
        return usable_soc * self.params.Q_capacity

File: web/packages/drivetrain-sim/src/components/battery.ts
/**
 * Battery component for hybrid and electric drivetrains.
 *
 * Features SOC-dependent OCV and internal resistance:
 * - Nonlinear OCV curve typical of Li-ion NMC chemistry
 * - Resistance increases at low and high SOC extremes
 *
 * References:
 * - A Generalized SOC-OCV Model for Lithium-Ion Batteries, Energies 2016
 */

import { DrivetrainComponent, PortValues } from '../core/component';
import { Port, PortDirection, PortType } from '../core/ports';
import { KinematicConstraint } from '../core/constraints';

/**
 * OCV model parameters for polynomial SOC-OCV relationship.
 */
export interface BatteryOcvParams {
  /** Polynomial coefficients for normalized OCV (V_oc / V_nom) */
  coefficients?: number[];
  /** Use lookup table instead of polynomial */
  useLookup?: boolean;
  /** SOC points for lookup */
  socPoints?: number[];
  /** Normalized OCV values for lookup */
  ocvNormalized?: number[];
}

/**
 * Resistance model parameters for SOC-dependent resistance.
 */
export interface BatteryResistanceParams {
  /** Nominal internal resistance [Ω] */
  rNom?: number;
  /** Low-SOC resistance increase factor */
  kLow?: number;
  /** Low-SOC decay constant */
  tauLow?: number;
  /** High-SOC resistance increase factor */
  kHigh?: number;
  /** High-SOC decay constant */
  tauHigh?: number;
}

/**
 * Battery parameters.
 */
export interface BatteryParams {
  /** Energy capacity [kWh] */
  capacityKwh?: number;
  /** Nominal open circuit voltage [V] at SOC=0.5 */
  vOc?: number;
  /** Nominal voltage [V] */
  vNom?: number;
  /** Nominal internal resistance [Ω] */
  rInt?: number;
  /** Maximum discharge power [W] */
  pMaxDischarge?: number;
  /** Maximum charge power [W] */
  pMaxCharge?: number;
  /** Minimum operating SOC */
  socMin?: number;
  /** Maximum operating SOC */
  socMax?: number;
  /** Initial SOC */
  socInit?: number;
  /** Enable SOC-dependent OCV model */
  useSocDependentOcv?: boolean;
  /** Enable SOC-dependent resistance model */
  useSocDependentResistance?: boolean;
  /** OCV model parameters */
  ocvParams?: BatteryOcvParams;
  /** Resistance model parameters */
  resistanceParams?: BatteryResistanceParams;
}

/** Default OCV model parameters (polynomial coefficients) */
const DEFAULT_OCV_PARAMS: Required<BatteryOcvParams> = {
  // Gives ~0.90 at SOC=0, ~1.0 at SOC=0.5, ~1.10 at SOC=1.0
  coefficients: [0.900, 0.400, -0.800, 1.200, -0.700, 0, 0, 0],
  useLookup: false,
  socPoints: [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
  ocvNormalized: [0.90, 0.92, 0.95, 0.97, 0.99, 1.00, 1.01, 1.03, 1.05, 1.08, 1.10],
};

/** Default resistance model parameters */
const DEFAULT_RESISTANCE_PARAMS: Required<BatteryResistanceParams> = {
  rNom: 0.05,
  kLow: 0.5,
  tauLow: 0.15,
  kHigh: 0.3,
  tauHigh: 0.15,
};

/** CAT 793D battery parameters */
export const CAT_793D_BATTERY_PARAMS: BatteryParams = {
  capacityKwh: 200.0,
  vOc: 750.0,
  vNom: 700.0,
  rInt: 0.05,
  pMaxDischarge: 1_000_000.0,
  pMaxCharge: 500_000.0,
  socMin: 0.3,
  socMax: 0.8,
  socInit: 0.6,
  useSocDependentOcv: true,
  useSocDependentResistance: true,
};

/** Larger battery for series hybrid */
export const SERIES_HYBRID_BATTERY_PARAMS: BatteryParams = {
  capacityKwh: 400.0,
  vOc: 750.0,
  vNom: 700.0,
  rInt: 0.03,
  pMaxDischarge: 1_500_000.0,
  pMaxCharge: 800_000.0,
  socMin: 0.2,
  socMax: 0.9,
  socInit: 0.6,
};

/** Large battery for pure EV */
export const EV_BATTERY_PARAMS: BatteryParams = {
  capacityKwh: 600.0,
  vOc: 800.0,
  vNom: 750.0,
  rInt: 0.02,
  pMaxDischarge: 2_000_000.0,
  pMaxCharge: 1_000_000.0,
  socMin: 0.1,
  socMax: 0.95,
  socInit: 0.8,
};

/**
 * Battery energy storage component.
 *
 * Uses equivalent circuit model with SOC-dependent parameters:
 * V_terminal = V_oc(SOC) - I × R(SOC)
 *
 * Power limits are derated near SOC boundaries to protect the battery.
 */
export class BatteryComponent extends DrivetrainComponent {
  readonly params: Required<Omit<BatteryParams, 'ocvParams' | 'resistanceParams'>> & {
    ocvParams: Required<BatteryOcvParams>;
    resistanceParams: Required<BatteryResistanceParams>;
  };
  private _soc: number;

  constructor(params: BatteryParams = {}, name: string = 'battery') {
    super(name);

    this.params = {
      capacityKwh: params.capacityKwh ?? 200.0,
      vOc: params.vOc ?? 750.0,
      vNom: params.vNom ?? 700.0,
      rInt: params.rInt ?? 0.05,
      pMaxDischarge: params.pMaxDischarge ?? 1_000_000.0,
      pMaxCharge: params.pMaxCharge ?? 500_000.0,
      socMin: params.socMin ?? 0.3,
      socMax: params.socMax ?? 0.8,
      socInit: params.socInit ?? 0.6,
      useSocDependentOcv: params.useSocDependentOcv ?? true,
      useSocDependentResistance: params.useSocDependentResistance ?? true,
      ocvParams: { ...DEFAULT_OCV_PARAMS, ...params.ocvParams },
      resistanceParams: { ...DEFAULT_RESISTANCE_PARAMS, ...params.resistanceParams },
    };

    this._soc = this.params.socInit;
  }

  /** Capacity in Joules. */
  get qCapacity(): number {
    return this.params.capacityKwh * 3600 * 1000;
  }

  /** Capacity in Amp-hours. */
  get qAh(): number {
    return this.qCapacity / (this.params.vNom * 3600);
  }

  /** Current state of charge [0-1]. */
  get soc(): number {
    return this._soc;
  }

  set soc(value: number) {
    this._soc = Math.max(0, Math.min(1, value));
  }

  get ports(): Record<string, Port> {
    return {
      electrical: {
        name: 'electrical',
        portType: PortType.ELECTRICAL,
        direction: PortDirection.BIDIRECTIONAL,
        description: 'Electrical power connection',
      },
    };
  }

  get stateNames(): string[] {
    return ['SOC'];
  }

  getInertia(_portName: string): number {
    return 0; // Battery has no mechanical inertia
  }

  getConstraints(): KinematicConstraint[] {
    return [];
  }

  /**
   * Get internal resistance at given SOC.
   *
   * R(SOC) = R_nom * (1 + k_low*exp(-SOC/tau_low) + k_high*exp((SOC-1)/tau_high))
   */
  getInternalResistance(soc?: number): number {
    const currentSoc = Math.max(0, Math.min(1, soc ?? this._soc));

    if (!this.params.useSocDependentResistance) {
      return this.params.rInt;
    }

    const rp = this.params.resistanceParams;
    const rLowFactor = rp.kLow * Math.exp(-currentSoc / rp.tauLow);
    const rHighFactor = rp.kHigh * Math.exp((currentSoc - 1) / rp.tauHigh);

    return rp.rNom * (1 + rLowFactor + rHighFactor);
  }

  /**
   * Get open circuit voltage at given SOC.
   *
   * Uses polynomial or lookup table model for SOC-OCV relationship.
   */
  getOpenCircuitVoltage(soc?: number): number {
    const currentSoc = Math.max(0, Math.min(1, soc ?? this._soc));

    if (!this.params.useSocDependentOcv) {
      return this.params.vOc;
    }

    const op = this.params.ocvParams;
    let ocvNorm: number;

    if (op.useLookup) {
      // Linear interpolation from lookup table
      ocvNorm = this.interpLookup(currentSoc, op.socPoints, op.ocvNormalized);
    } else {
      // Polynomial model: sum(a_i * SOC^i)
      ocvNorm = 0;
      for (let i = 0; i < op.coefficients.length; i++) {
        ocvNorm += op.coefficients[i] * Math.pow(currentSoc, i);
      }
    }

    return this.params.vOc * ocvNorm;
  }

  /** Linear interpolation helper */
  private interpLookup(x: number, xp: number[], yp: number[]): number {
    if (x <= xp[0]) return yp[0];
    if (x >= xp[xp.length - 1]) return yp[yp.length - 1];

    for (let i = 1; i < xp.length; i++) {
      if (x <= xp[i]) {
        const t = (x - xp[i - 1]) / (xp[i] - xp[i - 1]);
        return yp[i - 1] + t * (yp[i] - yp[i - 1]);
      }
    }
    return yp[yp.length - 1];
  }

  /**
   * Calculate current from power demand.
   *
   * Solves: P = V_oc × I - R_int × I²
   * Uses SOC-dependent OCV and internal resistance if enabled.
   *
   * @param power - Electrical power [W] (positive = discharge)
   * @returns Current [A] (positive = discharge)
   */
  getCurrentFromPower(power: number, soc?: number): number {
    const currentSoc = soc ?? this._soc;
    const vOc = this.getOpenCircuitVoltage(currentSoc);
    const R = this.getInternalResistance(currentSoc);

    if (Math.abs(power) < 1e-6) {
      return 0;
    }

    // Quadratic: R × I² - V_oc × I + P = 0
    // I = (V_oc - sqrt(V_oc² - 4RP)) / (2R)
    const discriminant = vOc * vOc - 4 * R * power;

    if (discriminant < 0) {
      // Power exceeds capability - return limiting current
      return power > 0 ? vOc / (2 * R) : -vOc / (2 * R);
    }

    // Use solution that gives smaller magnitude current
    return (vOc - Math.sqrt(discriminant)) / (2 * R);
  }

  /**
   * Get terminal voltage at given current.
   * Uses SOC-dependent OCV and internal resistance if enabled.
   */
  getTerminalVoltage(current: number, soc?: number): number {
    const currentSoc = soc ?? this._soc;
    const vOc = this.getOpenCircuitVoltage(currentSoc);
    const R = this.getInternalResistance(currentSoc);
    return vOc - current * R;
  }

  /**
   * Get power limits with SOC-based derating.
   *
   * @returns [P_min, P_max] [W]
   *   P_min is negative (max charging), P_max is positive (max discharging)
   */
  getPowerLimits(soc?: number): [number, number] {
    const currentSoc = soc ?? this._soc;

    let pDischarge = this.params.pMaxDischarge;
    let pCharge = this.params.pMaxCharge;

    // Derate discharge at low SOC
    if (currentSoc < this.params.socMin + 0.1) {
      const factor = Math.max(0, (currentSoc - this.params.socMin) / 0.1);
      pDischarge *= factor;
    }

    // Derate charge at high SOC
    if (currentSoc > this.params.socMax - 0.1) {
      const factor = Math.max(0, (this.params.socMax - currentSoc) / 0.1);
      pCharge *= factor;
    }

    return [-pCharge, pDischarge];
  }

  /**
   * Clip power to valid range.
   */
  clipPower(power: number, soc?: number): number {
    const [pMin, pMax] = this.getPowerLimits(soc);
    return Math.max(pMin, Math.min(power, pMax));
  }

  /**
   * Calculate SOC rate of change.
   *
   * dSOC/dt = -I / Q_capacity
   */
  getSocDerivative(power: number, soc?: number): number {
    const currentSoc = soc ?? this._soc;
    const current = this.getCurrentFromPower(power, currentSoc);
    // Q_capacity is in Coulombs (A·s)
    const Q = this.qCapacity / this.params.vNom;
    return -current / Q;
  }

  /**
   * Check if battery can provide requested power.
   */
  canProvidePower(power: number, soc?: number): boolean {
    const [pMin, pMax] = this.getPowerLimits(soc);
    return power >= pMin && power <= pMax;
  }

  /**
   * Get remaining usable energy [J].
   */
  getEnergyRemaining(soc?: number): number {
    const currentSoc = soc ?? this._soc;
    const usableSoc = Math.max(0, currentSoc - this.params.socMin);
    return usableSoc * this.qCapacity;
  }

  computeTorques(
    _portSpeeds: Record<string, number>,
    _controlInputs: Record<string, number>,
    _internalStates?: Record<string, number>
  ): Record<string, number> {
    return {}; // Battery doesn't produce torque
  }

  computeStateDerivatives(
    internalStates: Record<string, number>,
    portValues: PortValues
  ): Record<string, number> {
    const soc = internalStates.SOC ?? this._soc;
    const power = portValues.electrical_power ?? 0;

    const dSoc = this.getSocDerivative(power, soc);
    return { SOC: dSoc };
  }
}
